returning generic (?) function from main causes linker errors

Use this to reproduce the problem:
```
(defmodule NewSystem
  (deftemplate exit (Fn [Int] a) "$a $NAME(int code)" "$DECL { exit(code); }"))

(defn abort [error-code message]
  (do
    (IO.errorln message)
  error-code ))
(defn finish [message]
  (do
    (IO.println message)
  0 ))

(defn process [s1 s2]
  (if (String.= s1 s2)
    (abort 2 "come on, those parameters should not match!")
    (finish  "After lenghty computation we have reached a verdict...") ))

(defn main []
 ; (ignore (the Int
   (NewSystem.exit
    (let [ args (Int.- (StaticArray.length &System.args) 1) ]    ; 1st argument is executable!
    (if (not (Int.= 2 args))
      (abort 1 &(fmt "expecting exactly 2 arguments (got %d)." args))
      (process (StaticArray.unsafe-nth &System.args 1) (StaticArray.unsafe-nth &System.args 2)) ))))
  ; ))
 ```
 Here's the error output on Windows using gcc:
 ```
D:/.../x86_64-w64-mingw32/lib/../lib/libmingw32.a(lib64_libmingw32_a-crt0_c.o):crt0_c.c:(.text.startup+0x2e): undefined reference to `WinMain'
collect2.exe: error: ld returned 1 exit status
carp: callCommand: gcc  -o out\Untitled -I D:\Tools\p\carp/core/  -D_CRT_SECURE_NO_WARNINGS out\main.c (exit 1): failed
 ```

--- a/src/Commands.hs
+++ b/src/Commands.hs
@@ -342,7 +342,11 @@ commandBuild ctx [XObj (Bol shutUp) _ _] = do
         if generateOnly
           then pure (ctx, dynamicNil)
           else case Map.lookup "main" (envBindings env) of
-            Just _ -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ IntTy _)))) -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ UnitTy _)))) -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ _ _)))) -> pure (evalError ctx "`main` was generic at build time, it must return either `Int` or `()`." Nothing)
+            Just (Binder _ (XObj _ _ (Just t))) -> pure (evalError ctx ("`main` is expected to be a function, but was `" ++ show t ++ "`.") Nothing)
+            Just (Binder _ (XObj _ _ Nothing)) -> pure (evalError ctx ("The type of `main` is not set at build time. This should not happen, consider raising an issue with the code that caused it.") Nothing)
             Nothing -> compile False
   where
     removeSpecials env =

--- a/src/Commands.hs
+++ b/src/Commands.hs
@@ -344,6 +344,7 @@ commandBuild ctx [XObj (Bol shutUp) _ _] = do
           else case Map.lookup "main" (envBindings env) of
             Just (Binder _ (XObj _ _ (Just (FuncTy _ IntTy _)))) -> compile True
             Just (Binder _ (XObj _ _ (Just (FuncTy _ UnitTy _)))) -> compile True
+            -- we do not check for non-generic versions of main, these are already caught during concretization
             Just (Binder _ (XObj _ _ (Just (FuncTy _ _ _)))) -> pure (evalError ctx "`main` was generic at build time, it must return either `Int` or `()`." Nothing)
             Just (Binder _ (XObj _ _ (Just t))) -> pure (evalError ctx ("`main` is expected to be a function, but was `" ++ show t ++ "`.") Nothing)
             Just (Binder _ (XObj _ _ Nothing)) -> pure (evalError ctx ("The type of `main` is not set at build time. This should not happen, consider raising an issue with the code that caused it.") Nothing)

